design/Proposal-to-enhance-schema-for-metal3.md

```diff
@@ -0,0 +1,228 @@
+      - ram
+      - cpu
+      type: object
+    ````
+    
+    Where,
+    * `ram` a mapping from memory available to a `NUMA` node
+    * `cpus` a mapping from physical CPU ID to a `NUMA` node and a list of sibling threads
+    * `nics` a mapping from NIC names to `NUMA` node
+
+    Boot Parameters:
+    
+    * current_boot_mode: `uefi`
+        
+        UEFI is installed when device is manufactured. It is the first program which runs when the device is turned on. UEFI replace BIOS.
+    * pxe_interface: `ff:ff:ff:ff:ff:ff`
+        
+        The boot interface manages booting of both the deploy ramdisk and user interfaces on bare metal node. PXE boot interface is generic and works with all hardware that supports booting from network.
+    
+    Declarative design idea:
+    ````yaml
+    boot:
+        properties:
+            current_boot_mode:
+                type: string
+                description: current boot mode
+            pxe_interface:
+                type: string
+                description: pxe boot interface supports booting from network
+        type: object
+        required:
+        - current_boot_mode
+        - pxe_interface
+     ````
+2. Enhance struct in `baremetalhost_types.go` to store values for additional proposed parameters.
+    ````yaml
+    type CPU struct {
+    ...
+        NumaTopology NumaTopology `json:"numaTopology"`
+    }
+    
+    type NumaTopology struct {
+        NNics NumaNICS `json:"nNics"`
+        NRam  NumaRAM  `json:"nRam"`
+        NCPU  NumaCPU  `json:"nCpu"`
+    }
+    type NumaNICS struct {
+        NumaNodeId int    `json:"numaNodeID"`
+        Name       string `json:"name"`
+    }
+    type NumaRAM struct {
+        NumaNodeId int `json:"numaNodeID"`
+        SizeKb     int `json:"sizeKb"`
+    }
+    type NumaCPU struct {
+        NumaNodeId     int `json:"numaNodeID"`
+        CpuId          int `json:"cpuID"`
+        ThreadSiblings int `json:"threadSiblings"`
+    }
+    type Boot struct {
+        CurrentBootMode string `json:"currentBootMode"`
+        PxeInterface string `json:"pxeInterface"`
+    }
+    ````
+3. Changes in ironic.go:
+
+* To add support for parameters of `NUMA` topology we will update `getCPUDetails()` function.
+* Add new function `getBootDetails()` which will get details like `CurrentBootMode` and `PxeInterface` 
+and it will be called in `getHardwareDetails()` function.
+
+### Risks and Mitigations
+
+None
+
+## Design Details
+
+![Implementation-flow](enhance-schema-for-metal3-flow.png)
+
+*The changes are illustrated  in Implementation Details section.
+
+### Work Items
+
+1. Introduce new section called `boot` under `hardware` in schema `metal3.io_baremetalhosts_crd.yaml`.
+2. Update the existing sections for `cpu` which will include `NUMA` related parameters in schema.
+3. Define structure for `boot` parameters in `baremetalhosts_types.go`.
+4. Define structure for `NUMA` parameters and call it in `cpu` structure in `baremetalhost_types.go`.
```

Comment:
Does the gopher cloud library already return the NUMA data?

Reply:
No, We will have to add structure for NUMA in gophercloud lib. We will have to talk to gophercloud community for that or create a PR/issue for that. I welcome your help on that.

Reply:
No, We need to add NUMA structure in gopher library to support this feature. I welcome your help on that.

Reply:
That gophercloud task should be on the work list here.

Reply:
Yes, will add

Reply:
Numa details in gophercloud merged successfully 
https://github.com/gophercloud/gophercloud/pull/1842